### PR TITLE
Update remote host resources manager

### PR DESCRIPTION
**Pull request by lkolacek**
Signed-off-by: Ladislav Kolacek <lkolacek@redhat.com>

CLOUDBLD-8553

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] JSON/YAML configuration changes are updated in the relevant schema
- [x] Changes to metadata also update the documentation for the metadata
- [x] Pull request has a link to an osbs-docs PR for user documentation updates
- [x] New feature can be disabled from a configuration file

### CODE REVIEW

File: atomic_reactor/utils/remote_hosts_resources.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+
+    def _validate_resource_lock(self, payload: dict):
+        """ Validate resource lock json file against schema """
+        with open(RESOURCE_LOCK_SCHEMA, 'r') as f:
+            schema_data = f.read()
+        schema = json.loads(schema_data)
+        try:
+            jsonschema.validate(payload, schema)
+        except jsonschema.ValidationError as err:
+            logger.warning("Invalid json file: %s", err) #  TODO - What here? Fail build? Leave it be? Set file to empty json?
```
Comment (MartinBasti) [inline atomic_reactor/utils/remote_hosts_resources.py:82]: Don't fail build, just skip it

File: atomic_reactor/utils/resource_manager.py [hunk 2]
```diff
@@ -6,193 +6,290 @@
...
 from shlex import quote
+from typing import Any, Dict, List, Tuple
+
+SSH_COMMAND_TIMEOUT = 120
+RELATIVE_PATH = 'resource_manager'
+RESOURCE_INFO_JSON = 'info.json'
+RETRY_ON_SSH_EXCEPTIONS = (paramiko.ssh_exception.AuthenticationException,
+                           paramiko.ssh_exception.ChannelException,
+                           paramiko.ssh_exception.NoValidConnectionsError,
+                           paramiko.ssh_exception.SSHException)
```
Comment (MartinBasti) [inline atomic_reactor/utils/resource_manager.py:26]:
```
Subclasses:     AuthenticationException, ChannelException, BadHostKeyException, IncompatiblePeer, ProxyCommandFailure, CouldNotCanonicalize, ConfigParseError
```

I don't think we have to explicitly mention `AuthenticationException` and `ChannelException`

File: atomic_reactor/utils/resource_manager.py [hunk 3]
```diff
@@ -6,193 +6,290 @@
...
 import jsonschema
-from typing import List, Any, Dict, Tuple
+import paramiko
+import random
+import os
 from datetime import datetime
 from shlex import quote
+from typing import Any, Dict, List, Tuple
+
+SSH_COMMAND_TIMEOUT = 120
```
Comment (MartinBasti) [inline atomic_reactor/utils/resource_manager.py:20]: quite long?

File: atomic_reactor/utils/resource_manager.py [hunk 4]
```diff
@@ -6,193 +6,290 @@
...
-                logger.info('Locked')
-                return True
-            else:
-                logger.debug('Locking failed, retrying')
-                retries += 1
-        logger.info('Locking failed')
-        raise Exception('Locking failed')
+            client.connect(
+                self.hostname, username=self.username, key_filename=self.ssh_keyfile)
+        except paramiko.ssh_exception.AuthenticationException as ex:
```
Comment (MartinBasti) [inline atomic_reactor/utils/resource_manager.py:107]: Do we need this for each subclass?

File: atomic_reactor/utils/resource_manager.py [hunk 5]
```diff
@@ -6,193 +6,290 @@
...
+            logger.warning("SSH Connection error. %s", ex)
+        except paramiko.ssh_exception.SSHException as ex:
+            logger.warning("SSH Connection error. %s", ex)
+        return stdout, stderr, client.recv_exit_status()
+
+    def _ssh_run_remote_cmd_to_get_json(self, client, cmd: List):
+        stdin, stdout, stderr = self._ssh_run_remote_cmd(client, cmd)
+        try:
+            json_stdout = json.loads(stdout)
+        except ValueError as ex:  # Not valid JSON  #  TODO: Question - Do we want to fail build on loading invalid json from remote host????
```
Comment (MartinBasti) [inline atomic_reactor/utils/resource_manager.py:148]: Skip, don't fail build. We don't want to explode all builds because one slot/host is misconfigured

File: atomic_reactor/utils/resource_manager.py [hunk 6]
```diff
@@ -6,193 +6,290 @@
...
+                    logger.warning("%s - slot %s - json load failed", resource_info['hostname'], slot_id)
+                elif slot:  # Non-empty json file was loaded
+                    prid = slot.get('prid')
+                    logger.debug("%s - slot %s is occupied by %s", resource_info['hostname'], slot_id, prid)
+                    self._validate_resource_lock(slot)
+                elif not slot:  # Empty json file was loaded
+                    logger.debug("%s - slot %s is available", resource_info['hostname'], slot_id)
+                    available_slots.append(slot_id)
+                else:  # Undefined state - should not happen
+                    raise RuntimeError(f"{resource_info['hostname']} - slot {slot_id} invalid state!")
```
Comment (MartinBasti) [inline atomic_reactor/utils/resource_manager.py:187]: :eyes:

File: atomic_reactor/utils/resource_manager.py [hunk 7]
```diff
@@ -6,193 +6,290 @@
...
+
+        return random.shuffle(available_slots)
+
+    def get_cmd_to_read_from_json(self, file_name: str):
+        """ Create a command to retrieve content of json file from remote host """
+        return ['cat', os.path.join(RELATIVE_PATH, file_name)]
```
Comment (MartinBasti) [inline atomic_reactor/utils/resource_manager.py:193]: if you define RELATIVE_PATH as pathobject in module, you can just use
```python
str(RELATIVE_PATH / filename)
```

not sure if that's better

File: atomic_reactor/utils/resource_manager.py [hunk 8]
```diff
@@ -6,193 +6,290 @@
...
+        filepath = os.path.join(RELATIVE_PATH, file_name)
+        _cmd = (f'if grep -qw {{}} {filepath} ; then echo {quote(payload)} > '
+                f'{filepath} ;else exit 1; fi') #  TODO: Test with remote host that this is actually working with Paramiko
+        return f'flock {filepath} -c {quote(_cmd)}'
+
+    def lock(self, slot_id: int, prid: str):
+        logger.debug("Locking slot %s on host %s with pipelinerun ID %s",
+                     slot, self._hostname, prid)
+
+        payload = json.dumps({'prid': prid, 'locked': str(datetime.now())})
```
Comment (MartinBasti) [inline atomic_reactor/utils/resource_manager.py:210]: I'd prefer this here `datetime.utcnow().isoformat()`

File: atomic_reactor/utils/resource_manager.py [hunk 9]
```diff
@@ -6,193 +6,290 @@
...
+                                'returned non-zero exitcode')
+
+        logger.debug("%s - slot %s locked with pipelinerun ID %s", self._hostname, slot_id, prid)
+
+    def unlock(self, slot_id: int, prid: int):
+        logger.debug("Unlocking slot %s on host %s with pipelinerun ID %s",
+                     slot_id, self._hostname, prid)
+
+        with self._ssh_session() as client:
+            #  TODO: Is critical section needed here? -- Start
```
Comment (MartinBasti) [inline atomic_reactor/utils/resource_manager.py:227]: yes, but not part of this PR

File: atomic_reactor/utils/resource_manager.py [hunk 10]
```diff
@@ -6,193 +6,290 @@
...
+    @classmethod
+    def from_config(cls, config: dict):
+        """ Instantiate remote hosts loaded from configmap
+
+        :param config: dict, Arch specific remote hosts dictionary from configmap
+
+        Example:
+
+        hostname-remote-host1:
+          enabled: true
```
Comment (MartinBasti) [inline atomic_reactor/utils/resource_manager.py:267]: should this filter out not enabled hosts?

File: atomic_reactor/utils/resource_manager.py [hunk 11]
```diff
@@ -6,193 +6,290 @@
...
+                 for hostname, attr in config.items()]
+
+        return cls(hosts)
+
+    def lock_resource(self, prid: str) -> LockedResource:
+        resources = [
+            (host, host.available_slots())
+            for host in self.hosts
+        ]
+        resources.sorted(key=lambda x: len(x[1]), reverse=True)  # Sort based on number of available slots
```
Comment (MartinBasti) [inline atomic_reactor/utils/resource_manager.py:285]: bonus points for sorting based on `available/all` slots

File: atomic_reactor/utils/resource_manager.py [hunk 12]
```diff
@@ -6,193 +6,290 @@
...
+        Returns list of available slots
+
+        Slots should be randomized here, for smoother locking, because there will be
+        hunger games when every pipeline will try to lock slot with lowest number first
+        """
+        logger.debug("Retrieve list of available slots on host %s", self._hostname)
+
+        with self._ssh_session() as client:
+            cmd = self.get_cmd_to_read_from_json(RESOURCE_INFO_JSON)
+            resource_info = self._ssh_run_remote_cmd_to_get_json(client, cmd)
```
Comment (MartinBasti) [inline atomic_reactor/utils/resource_manager.py:170]: I'm thinking, shouldn't be this just single method?

File: atomic_reactor/utils/resource_manager.py [hunk 13]
```diff
@@ -6,193 +6,290 @@
...
+        except paramiko.ssh_exception.SSHException as ex:
+            logger.warning("SSH Connection error. %s", ex)
+        return client
 
     @backoff.on_exception(
         backoff.expo,
-        Exception,
-        factor=PARAMIKO_BACKOFF_FACTOR,
-        max_tries=PARAMIKO_MAX_RETRIES + 1,  # total tries is N retries + 1 initial attempt
+        RETRY_ON_SSH_EXCEPTIONS,
```
Comment (MartinBasti) [inline atomic_reactor/utils/resource_manager.py:119]: Don't we have quite nested retries here?

**Review by MartinBasti — CHANGES_REQUESTED** [review]

File: atomic_reactor/utils/resource_manager.py [hunk 14]
```diff
@@ -6,193 +6,290 @@
...
-logger = logging.getLogger(__name__)
+class LockedResource:
 
+    def __init__(host: RemoteHost, slot: int, prid: str):
+        """
+        :param slot: int, Remote host slot ID
+        :param prid: str, Pipeline run ID
+        """
+        self.slot = slot
+        self.prid = prid
```
Comment (MartinBasti) [inline atomic_reactor/utils/resource_manager.py:44]: and `self.host = host` ?

File: atomic_reactor/utils/resource_manager.py (hunk 10 above)
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:267]: depends if we want to do selection of enabled hosts here or on OSBS side. I am OK with selecting them here.

File: atomic_reactor/utils/resource_manager.py [hunk 15]
```diff
@@ -6,193 +6,290 @@
...
-        data = self._ssh_run_remote_command(cmd)
-        return data
 
-    def _read_slot_data(self, slot: int):
-        cmd = f"cat {ROOT_PATH}slot_{slot}.json"
-        data = self._ssh_run_remote_command(cmd)
-        return data
+class RemoteHost:
+
+    def __init__(self, hostname: str, username: str, ssh_keyfile: str):
```
Comment (MartinBasti) [inline atomic_reactor/utils/resource_manager.py:52]: There will be more attrs, so it's better to enforce params by name
```suggestion
    def __init__(self, *, hostname: str, username: str, ssh_keyfile: str):
```

File: atomic_reactor/utils/resource_manager.py (hunk 9 above)
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:227]: Ack

File: atomic_reactor/utils/resource_manager.py [hunk 16]
```diff
@@ -6,193 +6,290 @@
...
+                    prid = slot.get('prid')
+                    logger.debug("%s - slot %s is occupied by %s", resource_info['hostname'], slot_id, prid)
+                    self._validate_resource_lock(slot)
+                elif not slot:  # Empty json file was loaded
+                    logger.debug("%s - slot %s is available", resource_info['hostname'], slot_id)
+                    available_slots.append(slot_id)
+                else:  # Undefined state - should not happen
+                    raise RuntimeError(f"{resource_info['hostname']} - slot {slot_id} invalid state!")
+
+        return random.shuffle(available_slots)
```
[the remark this answers is not in the document]
Comment (MartinBasti) [reply atomic_reactor/utils/resource_manager.py:189]: Explanation why, may be handy (I know why, but will I know why in 3 months? :D)

[the remark this answers is not in the document]
Comment (MartinBasti) [reply atomic_reactor/utils/resource_manager.py:189]: Taking back, it's in docstring :|

File: atomic_reactor/utils/resource_manager.py [hunk 17]
```diff
@@ -6,193 +6,290 @@
...
+        with self._ssh_session() as client:
+            cmd = self.get_cmd_to_read_from_json(RESOURCE_INFO_JSON)
+            resource_info = self._ssh_run_remote_cmd_to_get_json(client, cmd)
+
+            available_slots = []
+            for slot_id in range(resource_info.get('max_slot_count', 1)):
+                cmd = self.get_cmd_to_read_from_json(f'slot_{slot_id}.json')
+                slot = self._ssh_run_remote_cmd_to_get_json(client, cmd)
+
+                if slot is None:  # Json load failed - file on remote host might be corrupted
```
Comment (MartinBasti) [inline atomic_reactor/utils/resource_manager.py:177]: How are slots created initially?

File: atomic_reactor/utils/resource_manager.py (hunk 12 above)
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:170]: Can you elaborate more here? do you mean dropping `_ssh_run_remote_cmd_to_get_json` and using only `_ssh_run_remote_cmd` with json load separately??

File: atomic_reactor/utils/resource_manager.py (hunk 3 above)
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:20]: 30? 60?

File: atomic_reactor/utils/resource_manager.py (hunk 17 above)
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:177]: I believe via our configuration playbook for remote hosts, but I'll check it in design.

File: atomic_reactor/utils/resource_manager.py (hunk 7 above)
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:193]: If there won't be any strong opinion on this, It will stay as it is now. :)

File: atomic_reactor/utils/resource_manager.py (hunk 4 above)
Comment (chmeliik) [reply atomic_reactor/utils/resource_manager.py:107]: Dosn't backoff log something by itself anyway?

Also note that if you suppress the exceptions, backoff will not work

File: atomic_reactor/utils/resource_manager.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (chmeliik) [reply atomic_reactor/utils/resource_manager.py:189]: Could we shuffle in lock_resource() instead? Trying the slots in a random order is relevant to the locking mechanism but not to this method.

And note that random.shuffle does `Shuffle list x in place, and return None.`

File: atomic_reactor/utils/resource_manager.py (hunk 11 above)
Comment (rcerven) [reply atomic_reactor/utils/resource_manager.py:285]: sorted is function not list method

File: atomic_reactor/utils/resource_manager.py [hunk 18]
```diff
@@ -6,193 +6,290 @@
...
+    def get_cmd_to_write_empty_slot(self, slot_id: int):
+        """ Create a command to unlock slot on remote host """
+        return ['echo', '{}', '>', os.path.join(RELATIVE_PATH, f'slot_{slot_id}.json')]
+
+    def get_cmd_to_write_slot_in_progress(self, file_name: str, payload: str):
+        """ Create a command to lock slot for pipelinerun on remote host """
+        filepath = os.path.join(RELATIVE_PATH, file_name)
+        _cmd = (f'if grep -qw {{}} {filepath} ; then echo {quote(payload)} > '
+                f'{filepath} ;else exit 1; fi') #  TODO: Test with remote host that this is actually working with Paramiko
+        return f'flock {filepath} -c {quote(_cmd)}'
```
Comment (rcerven) [inline atomic_reactor/utils/resource_manager.py:204]: this locks the file ONLY for duration of the command ......

you have to use file descriptor

1st shell:
$ exec 99>"lock.file"
$ flock -n 99
$ echo $?
0

2nd shell
$ exec 99>"lock.file"
$ flock -n 99
$ echo $?
1

1st shell:
$ flock -u -n 99

2nd shell 
$ flock -n 99
$ echo $?
0

but all that works ONLY if that shell is still active

Comment (rcerven) [reply atomic_reactor/utils/resource_manager.py:204]: also keep in mind that file actually isn't locked itself, any process can write to it

Comment (rcerven) [reply atomic_reactor/utils/resource_manager.py:204]: so basically what it does now is that it is using flock feature just for writing to the file .... (was that really the intent ?)

File: atomic_reactor/utils/resource_manager.py (hunk 4 above)
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:107]: Adam is right here. Logging can be added directly to backoff decorator and exception will break backoff completely. I made few tests here. https://gitlab.cee.redhat.com/-/snippets/4833 I'll make a fix.

File: atomic_reactor/utils/resource_manager.py [hunk 19]
```diff
@@ -6,193 +6,290 @@
...
         with open(RESOURCE_LOCK_SCHEMA, 'r') as f:
             schema_data = f.read()
         schema = json.loads(schema_data)
         try:
             jsonschema.validate(payload, schema)
-            return True
         except jsonschema.ValidationError as err:
-            logger.error(err)
-        return False
+            logger.warning("Invalid json file: %s", err) #  TODO - What here? Fail build? Leave it be? Set file to empty json?
```
[the remark this answers is not in the document]
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:82]: I would leave fixing the slot to cleanup job rather then doing it here. So I'll leave it as it is for now. Thanks for feedback.

File: atomic_reactor/utils/resource_manager.py [hunk 20]
```diff
@@ -6,193 +6,290 @@
...
-    free_host, slot = mngr.obtain_free_resource('unique-name-for-pipeline-run')
-
-    # Unlocking
-    free_host.unlock(slot, 'unique-name-for-pipeline-run')
+            stdin, stdout, stderr = client.exec_command(
+                cmd, timeout=SSH_COMMAND_TIMEOUT)
+        except paramiko.ssh_exception.AuthenticationException as ex:
+            logger.warning("SSH authentication failed. %s", ex) #  TODO: Question - Probably not needed for exec command
+        except paramiko.ssh_exception.ChannelException as ex:
+            logger.warning("Attempt to open a new Channel failed. %s", ex) #  TODO: Question - Probably not needed for exec command
```
[the remark this answers is not in the document]
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:137]: I agree, Tested here https://gitlab.cee.redhat.com/-/snippets/4833 I'll make a fix

File: atomic_reactor/utils/resource_manager.py (hunk 18 above)
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:204]: This part was not introduced by myself, I would like to re-test it to be sure it's working properly. But Anyway it's not complete, beacuse`critical section` will be required here for locking and unlocking as I mentioned in another comment, but it will be added in a separate PR.

File: atomic_reactor/utils/remote_hosts_resources.py (hunk 1 above)
Comment (lkolacek) [reply atomic_reactor/utils/remote_hosts_resources.py]: ack

File: atomic_reactor/utils/resource_manager.py (hunk 2 above)
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:26]: ok I'll remove those from list of exceptions

File: atomic_reactor/utils/resource_manager.py (hunk 3 above)
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:20]: I'll set this to 45 for now

File: atomic_reactor/utils/resource_manager.py (hunk 14 above)
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:44]: ack

File: atomic_reactor/utils/resource_manager.py (hunk 15 above)
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:52]: I agree

File: atomic_reactor/utils/resource_manager.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:189]: Ou I'll fix that. Thanks

File: atomic_reactor/utils/resource_manager.py (hunk 10 above)
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:267]: Ok enabled hosts will be selected here in resource_manager.

File: atomic_reactor/utils/resource_manager.py (hunk 11 above)
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py]: Ack. Fixed

File: atomic_reactor/utils/resource_manager.py (hunk 18 above)
Comment (chmeliik) [reply atomic_reactor/utils/resource_manager.py:180]: > so basically what it does now is that it is using flock feature just for writing to the file .... (was that really the intent ?)

Yes, I believe it was. `flock` is there only to prevent race conditions while locking and unlocking. The file does not have to stay flock-ed for the entire duration of the podman build

Comment (rcerven) [reply atomic_reactor/utils/resource_manager.py:180]: well that seems weird anyway such usage of flock

as I've said earlier, anything can still write to that file anyway

File: atomic_reactor/utils/resource_manager.py [hunk 21]
```diff
@@ -6,193 +6,304 @@
...
+
+        # Slots should be randomized here to avoid always locking the lowest numbers first
+        resources = []
+        for host in self.hosts:
+            available_slots = host.available_slots()
+            random.shuffle(available_slots)
+            resources.append((host, available_slots))
+
+        # Sort based on number of available slots
+        sorted(resources, key=lambda x: len(x[1]), reverse=True)
```
Comment (MartinBasti) [inline atomic_reactor/utils/resource_manager.py:296]: This?
```suggestion
        recources.sort(key=lambda x: len(x[1]), reverse=True)
```

File: atomic_reactor/utils/resource_manager.py [hunk 22]
```diff
@@ -6,193 +6,304 @@
...
+        # Add enabled clusters to list hosts
+        hosts = [RemoteHost(hostname=hostname, username=attr['username'], ssh_keyfile=attr['auth'])
+                 for hostname, attr in config.items() if attr['enabled']]
+
+        return cls(hosts)
+
+    @backoff.on_exception(
+        backoff.constant,
+        LockRetry,
+        max_time=TIMEOUT_FAIL_BUILD,
```
[the remark this answers is not in the document]
Comment (MartinBasti) [reply atomic_reactor/utils/resource_manager.py:282]: (optional, I'm not sure here)

File: atomic_reactor/utils/resource_manager.py [hunk 23]
```diff
@@ -6,193 +6,304 @@
...
-ROOT_PATH = 'resource_manager/'
 
-RESOURCE_LOCK_SCHEMA = 'schemas/resource-lock-content.json'
+class LockError(Exception):
+    """ Attempt for resource locking returned non-zero exitcode """
+
+
+class LockRetry(Exception):
+    """ Retry locking, no resources were available """
+
```
Comment (MartinBasti) [inline atomic_reactor/utils/resource_manager.py:41]: IMO we should inherit from RuntimeError (this is currently used in OSBS everywhere)
```suggestion
class ResourceManagerError(RuntimeError):
   """Base module exception"""

class LockError(ResourceManagerError):
    """ Attempt for resource locking returned non-zero exitcode """


class LockRetry(ResourceManagerError):
    """ Retry locking, no resources were available """

```

File: atomic_reactor/utils/resource_manager.py (hunk 13 above)
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:105]: I hope we should not now.

File: atomic_reactor/utils/resource_manager.py (hunk 12 above)
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:141]: Leaving as it is. one method is general for shell commands on remote host and second one is used for commands, which returns json format to be automatically loaded and passed as dict.

File: atomic_reactor/utils/resource_manager.py (hunk 17 above)
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:148]: Added note to ticket for follow up PR.

File: atomic_reactor/utils/resource_manager.py (hunk 23 above)
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py]: Added

File: atomic_reactor/utils/resource_manager.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:286]: @rcerven @chmeliik thoughts?

[the remark this answers is not in the document]
Comment (rcerven) [reply atomic_reactor/utils/resource_manager.py:286]: if locking fails with retries on 1 host we should move on to another candidate host

File: atomic_reactor/utils/resource_manager.py [hunk 24]
```diff
@@ -6,193 +6,308 @@
...
+RESOURCE_INFO_JSON = 'info.json'
+RETRY_ON_SSH_EXCEPTIONS = (paramiko.ssh_exception.NoValidConnectionsError,
+                           paramiko.ssh_exception.SSHException)
+BACKOFF_FACTOR = 5
+MAX_RETRIES = 3
 
-PARAMIKO_MAX_RETRIES = 3
-PARAMIKO_BACKOFF_FACTOR = 1
+logger = logging.getLogger('resource-manager')
+logger.setLevel(logging.ERROR)
```
Comment (chmeliik) [inline atomic_reactor/utils/resource_manager.py:32]:
```suggestion
logger = logging.getLogger(__name__)
```

File: atomic_reactor/utils/resource_manager.py [hunk 25]
```diff
@@ -6,193 +6,308 @@
...
+        return self._username
+
+    @property
+    def ssh_keyfile(self):
+        return self._ssh_keyfile
 
     def _validate_resource_lock(self, payload: dict):
+        """ Validate resource lock json file against schema """
         with open(RESOURCE_LOCK_SCHEMA, 'r') as f:
             schema_data = f.read()
```
Comment (chmeliik) [inline atomic_reactor/utils/resource_manager.py:74]: Will this work after installation? `RESOURCE_LOCK_SCHEMA` is a relative path. We should use `util.validate_with_schema` for this.

File: atomic_reactor/utils/resource_manager.py [hunk 26]
```diff
@@ -6,193 +6,308 @@
...
-            logger.error(err)
-        return False
+            logger.warning("Invalid json file: %s", err)
+
+    @contextmanager
+    def _ssh_session(self):
+        """ Create an SSH connection."""
+        client = self._open_ssh_session()
+        yield client
+        client.close()
```
Comment (chmeliik) [inline atomic_reactor/utils/resource_manager.py:86]: client.close() should be in `finally` block, otherwise connection won't be closed on exceptions

File: atomic_reactor/utils/resource_manager.py [hunk 27]
```diff
@@ -6,193 +6,308 @@
...
+    )
+    def _ssh_run_remote_cmd(self, client, cmd: List):
+        """
+        Run a Shell command on a remote host
 
-    def check_avail(self):
+        :return: stdout, stderr and exitcode of shell command
+        """
+        logger.debug("Try to run command %s on remote host %s", cmd, self.hostname)
+        _, stdout, stderr = client.exec_command(cmd, timeout=SSH_COMMAND_TIMEOUT)
```
Comment (chmeliik) [inline atomic_reactor/utils/resource_manager.py:122]: exec_command does not support passing the command as a list

we could convert it with `cmd_str =  " ".join(map(shlex.quote, cmd))`

File: atomic_reactor/utils/resource_manager.py [hunk 28]
```diff
@@ -6,193 +6,308 @@
...
+        return stdout, stderr, client.recv_exit_status()
```
Comment (chmeliik) [inline atomic_reactor/utils/resource_manager.py:123]:
```suggestion
        return stdout, stderr, stdout.channel.recv_exit_status()
```
The Client class does not have this method.

Also, according to [Paramiko docs](https://docs.paramiko.org/en/stable/api/channel.html#paramiko.channel.Channel.recv_exit_status), `recv_exit_status` may hang indefinitely unless you read from the output streams first. Maybe this method should fully consume stdout and stderr, and return strings rather than the open Channel Files.

File: atomic_reactor/utils/resource_manager.py [hunk 29]
```diff
@@ -6,193 +6,308 @@
...
-                retries += 1
-        logger.info('Locking failed')
-        raise Exception('Locking failed')
+            json_stdout = json.loads(stdout)
+        except ValueError as ex:  # Not valid JSON
+            logger.warning("Json load failed. Not a valid JSON. %s", ex)
+            return None
+        except TypeError as ex:  # Not an object
+            logger.warning("Json load failed. Not an object. %s", ex)
+            return None
```
Comment (chmeliik) [inline atomic_reactor/utils/resource_manager.py:134]: How would this even happen?

File: atomic_reactor/utils/resource_manager.py [hunk 30]
```diff
@@ -6,193 +6,308 @@
...
+                    logger.debug("%s - slot %s is occupied by %s",
+                                 resource_info['hostname'], slot_id, prid)
+                    self._validate_resource_lock(slot)
+                elif not slot:  # Empty json file was loaded
+                    logger.debug("%s - slot %s is available",
+                                 resource_info['hostname'], slot_id)
+                    available_slots.append(slot_id)
+                else:  # Undefined state - should not happen
+                    logger.warning("%s - slot %s invalid state!",
+                                   resource_info['hostname'], slot_id)
```
Comment (chmeliik) [inline atomic_reactor/utils/resource_manager.py:166]: Not just undefined state, straight up impossible unless the `__bool__` method of slots returns non-deterministic results :smile:

File: atomic_reactor/utils/resource_manager.py [hunk 31]
```diff
@@ -6,193 +6,308 @@
...
+
+        return available_slots
+
+    def get_cmd_to_read_from_json(self, file_name: str):
+        """ Create a command to retrieve content of json file from remote host """
+        return ['cat', os.path.join(RELATIVE_PATH, file_name)]
```
Comment (chmeliik) [inline atomic_reactor/utils/resource_manager.py:172]: IMO we should use an absolute path

File: atomic_reactor/utils/resource_manager.py [hunk 32]
```diff
@@ -6,193 +6,308 @@
...
+
+    def get_cmd_to_write_empty_slot(self, slot_id: int):
+        """ Create a command to unlock slot on remote host """
+        return ['echo', '{}', '>', os.path.join(RELATIVE_PATH, f'slot_{slot_id}.json')]
```
Comment (chmeliik) [inline atomic_reactor/utils/resource_manager.py:176]: Redirects are only interpreted by shells, `>` will not work in list form. But then again, `exec_command` does not support lists anyway.

File: atomic_reactor/utils/resource_manager.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (chmeliik) [reply atomic_reactor/utils/resource_manager.py:286]: That already happens inside this method. If *all* slots on *all* hosts fail to lock, that's when the method raises LockRetry and does the entire process again.

I don't mind having the extra timeout here.

**Review by chmeliik — COMMENTED** [review]
LGTM, the comments don't need to be addressed right away (but they do need to be addressed eventually)

**Review by chmeliik — APPROVED** [review]

File: atomic_reactor/utils/resource_manager.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (rcerven) [reply atomic_reactor/utils/resource_manager.py:286]: ic, well there are now 2 use cases:
1) something is wrong with hosts, that we are unable to lock any resource
2) there is more pipeline runs running than remote hosts capacity

for 2nd case we should wait until there is free slot

for 1st case mmm, there might be some temporarily network issue, or there might be some serious issue with hosts, so if we will wait indefinitely (pipeline run timeout will kill run, but it will take a lot time to discover issue)

File: atomic_reactor/utils/resource_manager.py (hunk 30 above)
Comment (lkolacek) [reply atomic_reactor/utils/resource_manager.py:166]: Yeah this should not ever happen :D .. I was not sure If I should add else or leave it be.

File: atomic_reactor/utils/resource_manager.py [hunk 33]
```diff
@@ -6,193 +6,309 @@
...
-PARAMIKO_MAX_RETRIES = 3
-PARAMIKO_BACKOFF_FACTOR = 1
-
-ROOT_PATH = 'resource_manager/'
+from typing import List
 
 RESOURCE_LOCK_SCHEMA = 'schemas/resource-lock-content.json'
+SSH_COMMAND_TIMEOUT = 45
+TIMEOUT_FAIL_BUILD = 3000
+RELATIVE_PATH = 'resource_manager'
```
Comment (tkdchen) [inline atomic_reactor/utils/resource_manager.py:24]: @lkolacek What is the relative path of this RELATIVE_PATH? Is the working directory changed at somewhere or is the working directory determined by the configured ssh login?

File: atomic_reactor/utils/resource_manager.py [hunk 34]
```diff
@@ -6,193 +6,309 @@
...
+        Run a Shell command on a remote host
+
+        :return: stdout, stderr and exitcode of shell command
+        """
+        logger.debug("Try to run command %s on remote host %s", cmd, self.hostname)
+        _, stdout, stderr = client.exec_command(cmd, timeout=SSH_COMMAND_TIMEOUT)
+        return stdout, stderr, stdout.channel.recv_exit_status()
 
-    def find(self) -> List[Host]:
+    def _ssh_run_remote_cmd_to_get_json(self, client, cmd: List):
```
Comment (tkdchen) [inline atomic_reactor/utils/resource_manager.py:126]: What is the json?

File: atomic_reactor/utils/resource_manager.py [hunk 35]
```diff
@@ -6,193 +6,309 @@
...
-
-    # Unlocking
-    free_host.unlock(slot, 'unique-name-for-pipeline-run')
+            json_stdout = json.loads(stdout)
+        except ValueError as ex:  # Not valid JSON
+            logger.warning("Json load failed. Not a valid JSON. %s", ex)
+            return None
+        except TypeError as ex:  # Not an object
+            logger.warning("Json load failed. Not an object. %s", ex)
+            return None
```
Comment (tkdchen) [inline atomic_reactor/utils/resource_manager.py:135]: @lkolacek What does the None mean to the caller?

- In `available_slots`,
  - None is not handled for the value of `resource_info`
  - None means "json load failed" for getting the value of `slot`.
- In `unlock`, None means a slot is available on a host and unlocking skipped. Looks like None is a correct return value as expected.
In

File: atomic_reactor/utils/resource_manager.py [hunk 36]
```diff
@@ -6,193 +6,309 @@
...
-    free_host, slot = mngr.obtain_free_resource('unique-name-for-pipeline-run')
-
-    # Unlocking
-    free_host.unlock(slot, 'unique-name-for-pipeline-run')
+            json_stdout = json.loads(stdout)
+        except ValueError as ex:  # Not valid JSON
+            logger.warning("Json load failed. Not a valid JSON. %s", ex)
+            return None
+        except TypeError as ex:  # Not an object
+            logger.warning("Json load failed. Not an object. %s", ex)
```
Comment (tkdchen) [inline atomic_reactor/utils/resource_manager.py:134]: Should we log more info about the invalid stdout? Perhaps we can log even the whole stdout value. That might be convenience for us to debug.

File: atomic_reactor/utils/resource_manager.py (hunk 30 above)
Comment (tkdchen) [reply atomic_reactor/utils/resource_manager.py:167]: I think the validation can be done earlier and continue the for-loop immediately when slot is None:

```
if slot is None:
    logger.warning()
    continue

self._validate_resource_lock(slot)

if slot:
    logger.debug("slot %s is occupied by %s")
else:
    logger.debug("slot is available.")
    available_slots.append(slot_id)
```

The data types are validated by the schema. Therefore, no "Undefined state" has to be cared about.

But, in either case, should the validation error be handled specifically? The `lock_resource` will fail if the validation fails. Then, will the pipeline run fail eventually? Or would it be good to log enough information for the invalid slot or the specific host and let the `lock_resource` locks other available resource as much as possible in order to keep the pipeline run?

Comment (chmeliik) [reply atomic_reactor/utils/resource_manager.py:167]: We discussed this a bit in one of the other threads - we want to log a warning and let the pipeline run try a different slot. Invalid slots could be fixed by the cleanup cronjob
